automatic formatted media data capability

chrisadubois opened the pull request:
slapped this together in a few hours today to get something in for the share client which needs data asap, we can add to it as we discover more about how we use it.

I tried to use what i had already created in the SDK for history/stats/analysis and that seemed to work to fulfill the requirement.

We could build this in the devices client pretty easily, but it seemed beneficial to add to the SDK to allow other 3rd party to use it since we are going to take the default :)

adds for share,audio,video -> flattens data to desired
automatically gathers based on config value

https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-117100

here's a sample output for share data that we can track on grafana for example
![Screen Shot 2020-01-28 at 4 50 41 PM](https://user-images.githubusercontent.com/12264933/73320438-28c44b00-41f4-11ea-99f2-edddc8a5addf.png)

Please review, thanks !

edit:

sample data printed in grafana
![Screen Shot 2020-01-30 at 12 11 44 PM](https://user-images.githubusercontent.com/12264933/73486871-c9347f80-435a-11ea-88ea-93ea332b137a.png)
![Screen Shot 2020-01-30 at 12 17 17 PM](https://user-images.githubusercontent.com/12264933/73486873-c9347f80-435a-11ea-8dbe-567b2190f4c7.png)

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/src/config.js:
```diff
@@ -25,8 +25,15 @@ export default {
+      historyMax: 120,
+      // should we move these to a separate config key?
+      // send a metrics event automatically
+      autoMediaMetrics: false,
```
I think we should rename this to something like `sendDiagnostics` or `autoUploadLogs` to make it more apparent to the end user what it does.

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/src/config.js:
```diff
@@ -25,8 +25,15 @@ export default {
+      // send a metrics event automatically
+      autoMediaMetrics: false,
+      // dispatch data every 10 seconds
+      mediaMetricsInterval: 10
```
Why is this different than the other interval? Seems like the naming could be tweaked here, too. Love all the configuration options though.

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/src/stats/data.js:
```diff
@@ -13,6 +16,28 @@ export default class WebRTCData {
+  flatten() {
+    let flat = {};
+
+    Object.keys(this.data).forEach((key) => {
```
Nit: I think `reduce` could accomplish the same thing without rewriting the flat variable in place.

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/src/stats/data.js:
```diff
@@ -13,6 +16,28 @@ export default class WebRTCData {
+  /**
+   * get the unique id for this specific stat pull
+   * @returns {Boolean};
+   */
```
Is this comment accurage? uniqueId on a bool?

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js:
```diff
@@ -0,0 +1,164 @@
+          id: STATS.videoReceiverId,
+          correlate: STATS.videoCorrelate,
+          metrics: true,
+          onData: () => {
```
I'm guessing we don't control the interval for getStats() so we have to do the check on every event message?

chrisadubois replied:
well, they are separate between getStats and the metrics needed for clients, and can be configured. getStats can be done entirely independently, so short answer yes.

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/src/stats/stats.js:
```diff
@@ -155,6 +167,20 @@ export default class MeetingStats extends StatelessWebexPlugin {
+   * @public
+   * @memberof MeetingStats
+   */
+  doMetrics(data) {
```
DO IT!

andrewholsted replied:
nit: `data.data` typically drives me nuts

chrisadubois replied:
i also hate it but it got messed up early on and now it's a decent sized refactor...

adamweeks commented on packages/node_modules/@webex/plugin-meetings/src/config.js:
```diff
@@ -25,8 +25,15 @@ export default {
-      historyMax: 1800
+      // hold the last 2 minute of a calls data
+      historyMax: 120,
+      // should we move these to a separate config key?
```
Please remove this comment.

adamweeks commented on packages/node_modules/@webex/plugin-meetings/src/constants.js:
```diff
@@ -293,6 +293,20 @@ export const DEFAULT_TRANSFORM_KEYS = [
   'rtcCandidatePairAudio',
   'rtcCandidatePairVideo'
 ];
+export const DEFAULT_FLATTEN_DATA_KEYS = [
```
Would be great to add a comment above this to indicate where these constants come from and are expected to be used.

adamweeks commented on packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:
```diff
@@ -983,6 +992,34 @@ export default class Meeting extends StatelessWebexPlugin {
+   * @public
+   * @memberof Meeting
+   */
+  getMediaMetrics() {
```
I know for consistency sake we are using getters around the meetings plugin, but was there a particular reason not to expose just the `this.mediaMetrics` property?

chrisadubois replied:
no particular reason, just consistency sake.

adamweeks commented on packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:
```diff
@@ -983,6 +992,34 @@ export default class Meeting extends StatelessWebexPlugin {
+   * @memberof Meeting
+   */
+  autoStats() {
+    const automatic = new StatsMetrics(this.attrs, this.options);
```
Renaming the var might make this a bit more readable:

```suggestion
    const automaticMetrics = new StatsMetrics(this.attrs, this.options);
```

adamweeks commented on packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js:
```diff
@@ -0,0 +1,164 @@
+
+  /**
+   * sets up all the default senders and receivers getStats to collect data
+   * @returns {undefined}
```
This is incorrect

adamweeks commented on packages/node_modules/@webex/plugin-meetings/src/stats/stats.js:
```diff
@@ -387,6 +438,10 @@ export default class MeetingStats extends StatelessWebexPlugin {
     return this.history;
   }
 
+  getMetrics() {
```
This needs jsdoc

chrisadubois commented:
@adamweeks we have zero tests around getStats currently -- for this PR to make it in, you'd like to me to add all of the following? unit, integration, and samples e2e tests -- I don't think all of these would be useful and most of them we can't do without media, I'd like to have a discussion on the ones we think are most important to get a mvp in for the feature.

arun3528 commented:
Were you able to find the bandwidth values?. i guess its under the transport

chrisadubois commented:
@arun3528 yes I found bandwidth values and added them to the report they are on the new candidatepair. so in this PR i also updated to match the newer fresher getStats API

added some unit tests, the getStats information which autoStats pulls from is already tested in the integration test and I don't want to slow it down another bit to wait for autoStats since they would be testing the same thing.

andrewholsted commented on packages/node_modules/@webex/plugin-meetings/src/stats/data.js:
```diff
@@ -13,6 +16,28 @@ export default class WebRTCData {
+  }
+
+  /**
+   * get the unique id for this specific stat pull
```
I still don't think this returns an id

adamweeks commented on packages/node_modules/@webex/plugin-meetings/src/constants.js:
```diff
@@ -764,6 +780,15 @@ export const SDP = {
 };
 
 export const STATS = {
+  audioCorrelate: 'audio',
```
Any reason why these are not all uppercase like the ones below it?

adamweeks commented on packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:
```diff
@@ -983,6 +992,34 @@ export default class Meeting extends StatelessWebexPlugin {
+   * Automatically publishes media metrics data and on a config interval
+   * uses the existing getStats data filter, so no new streams are created
+   * but it is separate than history, so history will not be available
+   * @returns {MediaMetrics}
```
This method doesn't `return`.

adamweeks commented on packages/node_modules/@webex/plugin-meetings/src/stats/stats.js:
```diff
@@ -155,6 +167,20 @@ export default class MeetingStats extends StatelessWebexPlugin {
+   */
+  doMetrics(data) {
+    if (this.metrics && data.data) {
+      if (!data.data.isEmpty()) {
```
`isEmpty` is a function on the `data.data` property?

adamweeks commented on packages/node_modules/@webex/plugin-meetings/test/integration/spec/journey.js:
```diff
@@ -44,6 +44,12 @@ skipInNode(describe)('plugin-meetings', () => {
     });
 
+    describe('Check configuration values', () => {
+      it('check config', () => {
```
```suggestion
      it('properly sets the `'autoSendMediaDiagnostics`' config', () => {
```

adamweeks commented on packages/node_modules/@webex/plugin-meetings/src/stats/transformer.js:
```diff
@@ -10,17 +10,19 @@ StatsTransformer.simplify = (stat) => find(keys(DEFAULT_TRANSFORM_REGEX), (key)
-    // some types need to check against a type field as well
-    if (DEFAULT_TRANSFORM_REGEX[value].profiler) {
-      const mapStatToReadableType = stat[DEFAULT_TRANSFORM_REGEX[value].profiler.type];
+    // TODO: wait a year to see if webrtc readds their different implementation for rtp/rtc/transport stats
```
Is there a link we could add to this comment to track the work being done on this, or the limitation?

adamweeks commented:
The merge commit needs to be replaced by a rebase on master and the tests are failing.

adamweeks commented:
Closing for now, @chrisadubois will reopen when ready.
